config/src/main/java/com/typesafe/config/ConfigDocument.java

```diff
@@ -0,0 +1,40 @@
+package com.typesafe.config;
+
+/**
+ * An object parsed from the original input text, which can be used to
+ * replace individual values and exactly render the original text of the
+ * input.
+ *
+ * <p>
+ * Because this object is immutable, it is safe to use from multiple threads and
+ * there's no need for "defensive copies."
+ *
+ * <p>
+ * <em>Do not implement interface {@code ConfigNode}</em>; it should only be
+ * implemented by the config library. Arbitrary implementations will not work
+ * because the library internals assume a specific concrete implementation.
+ * Also, this interface is likely to grow new methods over time, so third-party
+ * implementations will break.
+ */
+public interface ConfigDocument {
+    /**
+     * Returns a new ConfigDocument that is a copy of the current ConfigDocument,
+     * but with the desired value set at the desired path. If the path exists, it will
+     * remove all duplicates before the final occurrence of the path, and replace the value
+     * at the final occurrence of the path. If the path does not exist, it will be added.
+     *
+     * @param path the path at which to set the desired value
+     * @param newValue the value to set at the desired path, represented as a string. This
+     *                 string will be parsed into a ConfigNode, and the text will be inserted
+     *                 as-is into the document, with leading and trailing whitespace removed.
```

Comment:
In terms of whitespace trimming, this was primarily due to the fact that the desired value could be a simple value, which does not support multiple tokens/nodes. We could modify that for sure, though.

In terms of your question on value-setting: the ConfigDocument object saves the options which were used to parse it, and then uses those to parse any values passed to the `setValue()` method. If you parse a JSON file then set a value that uses HOCON syntax, it should conceivably throw a parse error. I don't have a test for that, but could easily add one.